### PR TITLE
Refactor handling of guns that don't consume ammo [RDY]

### PR DESCRIPTION
**Pull request by mugling**
Before this PR it was possible for an item to specify any combination of `ammo`, `NO_AMMO`, `BIO_WEAPON`, `RELOAD_AND_SHOOT`, `clip_size` and `magazine`. As a result there are a lot of corner cases such as that identified in #15606.
#### Behavioral changes

None anticipated
#### Implementation details
- Drops `NO_AMMO` replacing it with check for `ammo: "NULL"` (which becomes the default)
- Guns with `ammo` specify exactly one of `clip_size` or `magazine` and those without specify neither
- `RELOAD_AND_SHOOT` requires `ammo` to be set and prohibits setting of `magazine`
- `BIO_WEAPON` prohibits setting of `ammo` type
- Guns without `ammo` are normalized and moved to `data/items/gun`

| ammo | clip_size | magazine | RELOAD_AND_SHOOT | BIO_WEAPON |
| --- | --- | --- | --- | --- |
| yes | > 0 |  | optional |  |
| yes |  | yes |  |  |
| no |  |  |  | optional |

### CODE REVIEW

File: src/itype.h [hunk 1]
```diff
@@ -240,7 +240,7 @@ struct islot_gun : common_firing_data {
     /**
      * What type of ammo this gun uses.
      */
-    std::string ammo;
+    itype_id ammo = "NULL";
```
Comment (BevapDin) [inline src/itype.h]: Ammo isn't an item type, it's an ammo type. For example item types could be "arrow_metal" or "arrow_wood", but both have the ammo type "arrow". Those ids refer to different things. `ammotype` is the correct type to use here.

**Comment by Coolthulhu** [comment]
Produces a lot of warnings in mods: the tanks and other vehicles mod has a lot of guns that do not meet the stricter parser.

In general, it's a good idea to run `./cataclysm --check-mods` when changing the parser.

**Comment by mugling** [comment]
@Coolthulhu I'll look into that now

**Comment by mugling** [comment]
@Coolthulhu 06c27d8 should resolve that

**Comment by Coolthulhu** [comment]
Now all the turrets can be fired from hip, even though they're explicitly supposed to be usable only as vehicle weapons.

**Comment by mugling** [comment]
If I implement `VEHICLE_ONLY` as a separate PR?

**Comment by mugling** [comment]
> If I implement VEHICLE_ONLY as a separate PR?

Probably with associated descriptive text in `item::info`?

**Comment by mugling** [comment]
Considerably more explicit than relying upon `capacity: 0`

**Comment by Coolthulhu** [comment]
> If I implement VEHICLE_ONLY as a separate PR?

That would work

**Comment by kevingranade** [comment]
I'm still seeing json load errors from the generic guns mod. Basically it
overrides a number of guns back to the old integral magazine definition.
Ideally those definitions can be trimmed down to a targeted edit instead of
redefining the whole item.

**Comment by mugling** [comment]
@kevingranade yes, but I think that breakage is for a separate reason to this PR? I'm not especially familiar with the mod system but I'll take a look

**Comment by mugling** [comment]
@Coolthulhu `NO_RELOAD` suppresses both player reloading and display of capacity so I've gone with this in e242059 rather than introduce a new flag

File: data/json/items/ranged.json [hunk 2]
```diff
@@ -2705,7 +2445,7 @@
     "recoil": 0,
     "durability": 10,
     "burst": 5,
-    "clip_size": 0,
+    "clip_size": 100,
```
Comment (Coolthulhu) [inline data/json/items/ranged.json]: Lacks the flag, is reloadable now

File: data/mods/Tanks/items.json [hunk 3]
```diff
@@ -417,7 +417,7 @@
   "aim_speed" : 3,
   "recoil": -10,
   "durability": 9,
-  "clip_size": 0,
+  "clip_size": 1,
```
Comment (Coolthulhu) [inline data/mods/Tanks/items.json]: Hand-reloadable

**Comment by mugling** [comment]
@Coolthulhu both fixed in 1fb59ab

**Comment by Coolthulhu** [comment]
Semi-minor: trying to fire a non-reloadable weapon still gives "you have to reload" message.

**Comment by mugling** [comment]
Noted, I've begun work on reload messages starting with #15638

**Pull request merged**